helper/pool/pool.go

```diff
@@ -461,7 +496,7 @@ func (p *ConnPool) RPC(region string, addr net.Addr, version int, method string,
 			p.clearConn(conn)
 		}
 
-		p.releaseConn(conn)
+		conn.releaseUse()
```

Comment:
While reviewing this PR I'm noticing we have an existing hidden temporal coupling of `conn.releaseUse()` to `getRPCClient`. If someone were to call `conn.releaseUse` outside of this method without first calling `p.clearConn(conn)`, we can get a value of -1 for the reference count and then not close the connection. The current code is correct but easy for someone to break.

Two suggestions:
* Move this into a `defer conn.releaseUse()` right after we check the error from `getRPCClient` to make the temporal coupling more explicit.
* Maybe make `releaseUse` tolerant of misuse by having it close on `refCount < 1`:

```go
// releaseUse is the complement of `markForUse`, to free up the reference count
func (c *Conn) releaseUse() {
	refCount := atomic.AddInt32(&c.refCount, -1)
	if refCount < 1 && atomic.LoadInt32(&c.shouldClose) == 1 {
		c.Close()	
	}
}
```

Reply:
I agree it's brittle, I'll move to defer.

I'm unsure about handling negative refCounts. It's unclear how refCount can be negative by re-ordering calls. Only `conn.releaseUse` decrement the `refCount`. So negative values indicate double-release bugs, and defensive handling will probably lead to yet more subtle cases of unexpected connection closing like this one. I wish we can simply panic on negative values so we can find the source of double free case.

Reply:
> So negative values indicate double-release bugs, and defensive handling will probably lead to yet more subtle cases of unexpected connection closing like this one.

That's a good point. Totally agreed on that.